[Internal] Documentation: Adds msdata/direct Sync-up Guide. (#3828)
Azure/azure-cosmos-dotnet-v3: docs/sync_up_msdata_direct.md

# Update msdata/direct Repo with Cosmos v3 and Direct Codebase

## Table of Contents

> * [Backgraound.](#backgraound)

Reviewer: ```suggestion
* [Background.](#background)
```

Author: Thanks! Updated with latest commit.

* [Prerequisites.](#prerequisites)
* [Steps Required to Update msdata direct Repo.](#steps-required-to-update-msdata-direct-repo)
* [Validating the sync-up.](#validating-the-sync-up)
* [Submit Pull Request to msdata direct.](#submit-pull-request-to-msdata-direct)
* [Sample Pull Requests to Sync-up msdata direct.](#sample-pull-requests-to-sync-up-msdata-direct)

> ## Backgraound

Reviewer: ```suggestion
## Background
```

Author: Thanks! Updated with latest commit.

> As a developer on the Cosmos SDK team, we often engage in a task, that requires code changes in both cosmos dotnet sdk v3 repository, as well as in the `msdata` cosmosdb direct codebase (aka `Microsoft.Azure.Cosmos.Direct` namespace). Therefore, sometimes it's utter challanging to visualize the code changes as a whole, and analyze the impacts. To overcome this, we have created a branch called `msdata/direct` within our cosmos dotnet sdk v3 codebase, that basically mimics the code present in `msdata` repository mentioned above. This simplifies the code changes required to be done in both the repo and also provides much better understanding on the overall impacts of the code changes.

Reviewer: ```suggestion
As a developer on the Cosmos SDK team, we often engage in a task, that requires code changes in both cosmos dotnet sdk v3 repository, as well as in the `msdata` cosmosdb direct codebase (aka `Microsoft.Azure.Cosmos.Direct` namespace). Therefore, sometimes it's utterly challenging to visualize the code changes as a whole, and analyze the impacts. To overcome this, we have created a branch called `msdata/direct` within our cosmos dotnet sdk v3 codebase, that basically mimics the code present in `msdata` repository mentioned above. This simplifies the code changes required to be done in both, provides much better understanding on the overall impacts of the code changes, and gives transparency to users in terms of what is the source code for the `Microsoft.Azure.Cosmos.Direct` package.
```

Author: Thanks! Updated with latest commit.

## Prerequisites

Before covering the sync-up process in detail, please follow the below steps to make sure all the required pre-requisites are met.

### Clone the Azure Cosmos DB .NET SDK Version 3 Repo

- Clone the azure `cosmos-db dotnet sdk` repo in the local environment, using the below git command:
    -       git clone https://github.com/Azure/azure-cosmos-dotnet-v3.git

- Navigate to the directory `azure-cosmos-dotnet-v3` and check out the following branch, `msdata/direct` using the below git commands:
    -       git pull && git checkout msdata/direct

> ### Clone the `CosmosDB` Repo hosted in msdata

Reviewer: Cannot this section be simplified by just asking the person to follow the Cosmos DB onboarding steps to access the required repository without explicitly publicizing the repository url?

Author: Thanks Matias! This is a good point. I have updated the doc accordingly. Please feel free to take a look.

> - Setup the CosmosDB repository in the local environment, using the [CosmosDB onboarding guide](https://eng.ms/docs/cloud-ai-platform/azure-data/azure-data-azure-databases/cosmos-db-and-postgresql/azure-cosmos-db/azure-cosmos-db-onboarding-guide/engineering/dev-getting-started/clone-and-build-repo). Please note, building the entire repository is not required for the sync-up process.

Reviewer: Should we sharing publicly this internal link? Or can we simply say to follow the onboarding process to clone and build msdata enough? @kirankumarkolli any concerns with this link?

Reviewer 2: hmm unsure but seems safe to assume not to have the link

Author: I have updated the `md` file to remove the link.

> - Navigate to the clonned `CosmosDB` directory and check out the following branch, `master` using the below git commands:

Reviewer: ```suggestion
- Navigate to the cloned `CosmosDB` directory and check out the following branch, `master` using the below git commands:
```

Author: Thanks! Addressed!

- Navigate to the directory `CosmosDB` and check out the following branch, `master` using the below git commands:
    -       git pull && git checkout master

## Steps Required to Update msdata direct Repo

### Create a Feature Branch for Local Changes.

The first step to sync up the `msdata/direct` repo is to create a feature branch out of it, where all the required changes could be made. Later on, we will use the feature branch to submit pull request to `msdata/direct`. Please use the following git command to create the feature branch:

- Stay on the `msdata/direct` branch and run `git checkout -b users/<user_name>/update_msdata_direct_<mm_dd_yyyy>` to create the feature branch.

### Merging the cosmos db v3 Code into Feature Branch.

The next step is to port the latest `master` branch code into the newly created feature branch. Please see the below git commands to perform this action:

> - Stay on the newly created feature branch `users/<user_name>/update_msdata_direct_<mm_dd_yyyy>` and run `git merge master`. Make sure the `master` branch is up-to-date.

Reviewer: "Make sure master is up-to-date" should happen before `git merge master` :)

Author: Thanks! Updated with latest commit.

> - There are likely to be conflicts during the merge. If that happens, we will need to resolve the conflicts gracfully.

Reviewer: ```suggestion
- There are likely to be conflicts during the merge. If that happens, we will need to resolve the conflicts gracefully.
```

Author: Thanks! Updated with latest commit.

### Pick the Required Microsoft Azure Cosmos.Direct files into `msdata/direct` repo.

This is the last part for the sync-up process. Please follow the below steps to copy the required `Microsoft.Cosmos.Direct` files from msdata CosmosDB repo.

> - Open command prompt/windows terminal and navigate to the following directory where the cosmos v3 direct code is located, for example: `C:\stash\azure-cosmos-dotnet-v3\Microsoft.Azure.Cosmos\src\direct`.

Reviewer: This is on your machine right? Can we simplify to "Navigate to the `Microsoft.Azure.Cosmos\src\direct` folder in your cloned GitHub repo folder"?

Author: Thanks! Updated with latest commit.

The next step is to port the latest `master` branch code into the newly created feature branch. Please see the below git commands to perform this action:

- Make sure the `master` branch is up-to-date.
- Stay on the newly created feature branch `users/<user_name>/update_msdata_direct_<mm_dd_yyyy>` and run `git merge master`.
- There are likely to be conflicts during the merge. If that happens, we will need to resolve the conflicts gracefully by accepting the incoming `master` branch changes.

### Pick the Required Microsoft Azure Cosmos.Direct files into `msdata/direct` repo.

This is the last part for the sync-up process. Please follow the below steps to copy the required `Microsoft.Cosmos.Direct` files from msdata CosmosDB repo.

> - Open command prompt/windows terminal and navigate to the following directory `Microsoft.Azure.Cosmos\src\direct` inside the clonned cosmos v3 repo.

Reviewer: ```suggestion
- Open command prompt/windows terminal and navigate to the following directory `Microsoft.Azure.Cosmos\src\direct` inside the cloned cosmos v3 repo.
```

Author: Thanks! Addressed!
